### PR TITLE
✨ GraphQL reference docs

### PR DESCRIPTION
**Pull request by olyism**
New feature that generates markdown pages from a GraphQL API schema JSON export.

Here's a 5 min Loom video of me talking about the high-level UX and changes https://www.loom.com/share/d07d7374b4d2451e83c08750ded9fded

## Overview of changes

- [x] Set up script to generate all markdown pages
- [x] Render schema information onto pages
- [x] Navigation: dedicated nav for API, Integrations and GraphQL API docs
- [x] Styling 💄 💅 ✨
- [x] Update notification message
- [x] Linting fixes
- [x] Algolia to not index the schema pages
- [x] ~Better surface mutation~ Not needed as this is in query
- [x] Sort alphabetically
- [x] Add instructions

## Screenshots

GraphQL API's landing page with left nav divided by schema types.

Screen Shot 2022-07-18 at 9 01 20 pm

A schema page generated by the script.

Screen Shot 2022-07-18 at 9 03 15 pm

### CODE REVIEW

**Comment by olyism** [comment]
## Manual testing

Attaching screenshots of small and large screen sizes

### Chrome macOS

Screen Shot 2022-07-19 at 12 59 02 am

Screen Shot 2022-07-19 at 12 59 23 am

Screen Shot 2022-07-19 at 12 59 52 am

### Safari macOS

Screen Shot 2022-07-19 at 1 00 25 am

Screen Shot 2022-07-19 at 1 01 08 am

Screen Shot 2022-07-19 at 1 00 44 am

### Firefo macOS

Screen Shot 2022-07-19 at 1 02 53 am

Screen Shot 2022-07-19 at 1 02 32 am

Screen Shot 2022-07-19 at 1 03 18 am

**Comment by olyism** [comment]
Ok, so I've done some git squashing and condensed down to 14 commits. If you ignore the 🚀 generate pages commits then it's really 11 commits to look at.

Hopefully, this makes it easier to review the PR in a step-by-step manner. Please let me know if you need a walk through

**Comment by olyism** [comment]
Moved back to draft to make more fixes.

**Review by ddbeck — DISMISSED** [review]
@olyism I'm taking a look at the linting issues here (only—no comments on the substance of this) and I've got questions. Most importantly, where does the `graphql_data_schema.json` content come from? I'm guessing it's being drawn in from something external to the docs repo. To help with the linter problems, I think three things need to happen:

1. The pages generated by this process should be exempt from the linting rules by configuration (until they can be fixed upstream), not in-line annotations. More on this in a moment.
2. This generated content also needs to have a script to generate the `graphql_data_schema.json` data (i.e., there should be a well-structured way of getting the latest GraphQL schema). I'd say this ought to happen as part of this PR—I suspect you've already done _something_ to get it in the first place that could be included in the PR.
3. The problems reported by the linter ought to be fixed upstream, in the origin of `graphql_data_schema.json`.

On point 1, I'd request that you revert 93ea727 append something like this to the end of `vale.ini`:

```
## TODO: fix the issues in the underlying source for GraphQL schema content, then remove this configuration block
[*/pages/apis/graphql/schemas/*.md]
BasedOnStyles = Buildkite
Buildkite.spelling = NO
Buildkite.via = NO
Buildkite.ie = NO
```

A similar approach is available for `alex` and the other linting: we should skip directories known to cause trouble, rather than opt the files out one-by-one.

(And naturally, by making this comment, I'm volunteering myself to help fix issues in the underlying content.)

**Comment by plaindocs** [comment]
As an addition to your comment @ddbeck, I'm not sure I agree that generated files should be exempt. As you say, if there are upstream issues like the ones I raised in https://github.com/buildkite/buildkite/pull/9425 sure, but linting the markdown that we generate does help keep it it from being malformed.

**Comment by ddbeck** [comment]
@plaindocs To clarify, I meant more that we ought to exempt them from linting by configuration _until we've resolved the underlying issues that are creating the errors_, not that there should be a longstanding exemption for generated docs.

**Comment by olyism** [comment]
Screen Shot 2022-07-21 at 10 07 02 am

neat! Fixed the markdown linting trailing space error

**Comment by plaindocs** [comment]
Excellent work on fixing the linting here! <3

**Comment by olyism** [comment]
Have just edited Algolia config to exclude it from indexing GraphQL API docs' headings.

**Comment by plaindocs** [comment]
Awesome, I'll poke at the Algolia index, and see if it truely is ignored, and see if I can some eyes on this.

**Comment by plaindocs** [comment]
Confirm that the exclude syntax looks good. (CC @lachlanhardy )

**Comment by lachlanhardy** [comment]
@plaindocs Seems correct to me! Have you been able to rebuild the test index to confirm?

**Comment by olyism** [comment]
Just did some rebase, squashed some commits and force pushed to resolve conflicts and reduce number of commits.

**Comment by plaindocs** [comment]
Yep, the GraphQL stuff looks good - ie not indexed.

I reckon as a minimum we could leave the H1 **in** the index. What do folks think?

I don't super like `Schema browser` not being clickable. 
image

We could either write a quick overview page there, or just make it click to collapse? 

Also got some width issues on FF on Linux
image

File: pages/apis/graphql/schemas/agentconnection.md.erb [hunk 1]
```diff
@@ -0,0 +1,36 @@
+<!--
+  _____   ____    _   _  ____ _______   ______ _____ _____ _______ 
+  |  __  / __   |  | |/ __ __   __| |  ____|  __ _   _|__   __|
+  | |  | | |  | | |  | | |  | | | |    | |__  | |  | || |    | |   
+  | |  | | |  | | | . ` | |  | | | |    |  __| | |  | || |    | |   
+  | |__| | |__| | | |  | |__| | | |    | |____| |__| || |_   | |   
+  |_____/ ____/  |_| _|____/  |_|    |______|_____/_____|  |_|   
+  This file is auto-generated by script/generate_graphql_api_content.sh,
+  please build the schema.json by running `rails api:graph:export`
```
Comment (plaindocs) [inline pages/apis/graphql/schemas/agentconnection.md.erb:9]: Is this `rake` or `rails`? And what are the dependencies? More than the standard `buildkite/buildkite` ones?

**Comment by olyism** [comment]
> Yep, the GraphQL stuff looks good - ie not indexed.

Woot! That's great 🙌 

> I reckon as a minimum we could leave the H1 in the index. What do folks think?

I'm thinking leaving H1s out of the index for now. There are 380+ pages so I don't know what the impact will be on other pages. I would like to create a follow up ticket to iterate on the search experience somehow, so we can make it faster to find the schema.

> I don't super like Schema browser not being clickable.

I was tossing between the collapsed Schema Browser v.s. start expanded and decided to do the expanded option to reduce the number of nav levels. The un-clickable Schema Browser heading is not ideal, as it doesn't behave the same was as other links, but I'm wary of bloating this PR too much. If we're okay with this UX quirk, I'm happy to follow up with a PR for a Schema Browser index page over the next few days.

> Also got some width issues on FF on Linux

This is a word-break CSS behavior I added to prevent horizontal scrolling in the sidebar. It's a bit odd but less odd than horizontal scroll. I'd be keen to look into a more holistic UX fix to address very long query names in a later PR.

> I don't think any of these are blockers though.

If everyone's ok with the state of this PR, it'd be good to get it live to keep the PR small. If you feel strongly about a fix let me know also. Thanks :)

**Comment by yob** [comment]
wow, an epic PR! I love that graphql is going to get such great billing in the docs - the minimal approach we've had for a while was due for an uplift.

I haven't had a deep look at the diff yet, but after a skim my initial question is around the anticipated process for managing this over time. Will someone occasionally regenerate the schema file from the relevant codebase, import it to this repo, and re-generate the templates? Is that kind of similar to how the embedded agent docs are updated from time to time already?

**Comment by karensawrey** [comment]
This is going to be a legendary PR! I promise to have a deep long, thorough look at it tonight! 👍

**Comment by olyism** [comment]
hi @yob thanks for looking at this PR :)

> but after a skim my initial question is around the anticipated process for managing this over time

For the time being it'll be a manual process so, yes, it'll be someone generating the schema JSON and then running the script to generate the pages. We'll look out for changes to keep this section up to date.

**Comment by plaindocs** [comment]
@olyism yeah, that's what I mean by none of those comments being blockers. 

@yob I'm hoping we'll be able to properly automate both the Agent docs and this GraphQL schema in the not too distant future.

> be good to get it live to keep the PR small. 

sure mate. totally small. ;-)

**Review by plaindocs — APPROVED** [review]
Go for it.

**Comment by olyism** [comment]
Thanks very much everyone for your collaboration and support in this PR 🙌  will be merging soon and do some manual checks once it's in prod.

Once deployed and tested, I'll be raising several **much smaller** tickets to address the improvements raised in this thread.

File: pages/apis/graphql/schemas/__directive.md.erb [hunk 2]
```diff
@@ -1,3 +1,17 @@
+<!--
+  _____   ____    _   _  ____ _______   ______ _____ _____ _______ 
+  |  __  / __   |  | |/ __ __   __| |  ____|  __ _   _|__   __|
+  | |  | | |  | | |  | | |  | | | |    | |__  | |  | || |    | |   
+  | |  | | |  | | | . ` | |  | | | |    |  __| | |  | || |    | |   
+  | |__| | |__| | | |  | |__| | | |    | |____| |__| || |_   | |   
+  |_____/ ____/  |_| _|____/  |_|    |______|_____/_____|  |_|   
+  This file is auto-generated by script/generate_graphql_api_content.sh,
+  please build the schema.json by running `rails api:graph:export`
+  with https://github.com/buildkite/buildkite/,
```
Comment (jfhinchcliffe) [inline pages/apis/graphql/schemas/__directive.md.erb:10]: `s/with/within`?

**Comment by jfhinchcliffe** [comment]
@olyism this is amazing work! I don't know much about the docs repo, but does excluding it from algolia mean I can't search directly for a mutation / query that I'm after?

**Comment by olyism** [comment]
thanks @jfhinchcliffe 🙌 

> but does excluding it from algolia mean I can't search directly for a mutation / query that I'm after

Yes, that's correct. For the first cut we've decided to exclude it from Algolia for now so it doesn't risk inundating existing results (there are ~380 GraphQL API schema pages).

We'll have a separate ticket to investigate improving the search experience.

**Pull request merged**